Mathlib/NumberTheory/SelbergSieve.lean

```diff
@@ -0,0 +1,240 @@
+
+theorem prod_primeFactors_nu {d : ℕ} (hd : d ∣ P) : ∏ p ∈ d.primeFactors, ν p = ν d := by
+  rw [← nu_mult.map_prod_of_subset_primeFactors _ _ subset_rfl,
+    Nat.prod_primeFactors_of_squarefree <| Squarefree.squarefree_of_dvd hd prodPrimes_squarefree]
+
+theorem nu_pos_of_dvd_prodPrimes {d : ℕ} (hd : d ∣ P) : 0 < ν d := by
+  calc
+    0 < ∏ p ∈ d.primeFactors, ν p := by
+      apply prod_pos
+      intro p hpd
+      have hp_prime : p.Prime := prime_of_mem_primeFactors hpd
+      have hp_dvd : p ∣ P := (dvd_of_mem_primeFactors hpd).trans hd
+      exact nu_pos_of_prime p hp_prime hp_dvd
+    _ = ν d := prod_primeFactors_nu hd
+
+theorem nu_ne_zero {d : ℕ} (hd : d ∣ P) : ν d ≠ 0 := by
+  apply _root_.ne_of_gt
+  exact nu_pos_of_dvd_prodPrimes hd
+
+theorem nu_lt_one_of_dvd_prodPrimes {d : ℕ} (hdP : d ∣ P) (hd_ne_one : d ≠ 1) : ν d < 1 := by
+  have hd_sq : Squarefree d := Squarefree.squarefree_of_dvd hdP prodPrimes_squarefree
+  have := hd_sq.ne_zero
+  calc
+    ν d = ∏ p ∈ d.primeFactors, ν p := (prod_primeFactors_nu hdP).symm
+    _ < ∏ p ∈ d.primeFactors, 1 := by
+      apply prod_lt_prod_of_nonempty
+      · intro p hp
+        simp only [mem_primeFactors] at hp
+        apply nu_pos_of_prime p hp.1 (hp.2.1.trans hdP)
+      · intro p hpd; rw [mem_primeFactors_of_ne_zero hd_sq.ne_zero] at hpd
+        apply nu_lt_one_of_prime p hpd.left (hpd.2.trans hdP)
+      · simp only [nonempty_primeFactors, show 1 < d by omega]
+    _ = 1 := by
+      simp
+
+/-- The weight of all the elements that are a multiple of `d`. -/
+@[simp]
+def multSum (d : ℕ) : ℝ := ∑ n ∈ A, if d ∣ n then a n else 0
+
+@[inherit_doc multSum]
+scoped [SelbergSieve.Notation] notation3 "𝒜" => multSum
+
+/-- The remainder term in the approximation A_d = ν (d) X + R_d. This is the degree to which `nu`
+  fails to approximate the proportion of the weight that is a multiple of `d`. -/
+@[simp]
+def rem (d : ℕ) : ℝ := 𝒜 d - ν d * X
+
+@[inherit_doc rem]
+scoped [SelbergSieve.Notation] notation3 "R" => rem
+
+/-- The weight of all the elements that are not a multiples of any of our finite set of primes. -/
+def siftedSum : ℝ := ∑ d ∈ A, if Coprime P d then a d else 0
+
+/-- `X * mainSum μ⁺` is the main term in the upper bound on `sifted_sum`-/
+def mainSum (muPlus : ℕ → ℝ) : ℝ := ∑ d ∈ divisors P, muPlus d * ν d
+
+/-- `errSum μ⁺` is the error term in the upper bound on `sifted_sum`. -/
+def errSum (muPlus : ℕ → ℝ) : ℝ := ∑ d ∈ divisors P, |muPlus d| * |R d|
+
+theorem multSum_eq_main_err (d : ℕ) : multSum d = ν d * X + R d := by
+  dsimp [rem]
+  ring
+
+theorem siftedsum_eq_sum_support_mul_ite :
+    siftedSum = ∑ d ∈ support, a d * if Nat.gcd P d = 1 then 1 else 0 := by
+  dsimp only [siftedSum]
+  simp_rw [mul_ite, mul_one, mul_zero]
+
+omit s in
+/-- A sequence of coefficients $\mu^{+}$ is upper Moebius if $\mu * \zeta ≤ \mu^{+} * \zeta$. These
+  coefficients then yield an upper bound on the sifted sum.-/
+def UpperMoebius (muPlus : ℕ → ℝ) : Prop :=
```

Comment:
Mathlib seems to be converging towards a naming convention where predicates start with `Is` (or `Has`)
```suggestion
def IsUpperMoebius (muPlus : ℕ → ℝ) : Prop :=
```